### DISCUSSION TITLE
Where to instantiate a subscription that should be kept private for each user?

### DISCUSSION THREAD

**Discussion by scalemaildev** [socketio/socket.io#4114] [Q&A] [answered]
On an app I'm working on, I need each user to instantiate a server-side subscription that uses some critical account information. I'd like to keep each user's subscription (and its "serverClient") completely private to their connection. I'd prefer if it was also inaccessible to other tabs within their browser.

The current code for it looks something like this:
```
async function subscribeToTopic(io, topicIdString) {
        let sub = new topicQuery()
            .setTopicId(topicId)
            .subscribe(serverClient, res => {
                io.emit('newMessage', res);
            });
}
```

Both the "serverClient" and the actual subscription should be kept private to the user. So where would be the best place to instantiate this?

After reading through the docs I see that there's some functionality in Rooms for each connections "default room" that's automatically accessed by that socket. Would this be the correct place for the subscription? If so, how can I instantiate the serverClient so that it's only accessible from within that room? I'll also need to teardown the subscription (and clear the serverClient) once the user disconnects.

Thanks for any pointers.

**Comment by darrachequesne** [comment, marked as answer]
Hi! Topic subscriptions can easily be implemented with rooms:

```js
io.on("connection", (socket) => {
  // from the server side
  socket.join("the topic");

  // or from the client side
  socket.on("new subscription", (topic) => {
    socket.join(topic);
  });

  // and then
  io.to("the topic").emit("new message");
});
```

Though I'm not sure what you mean by "private to their connection". You could store the subscription on the socket object:

```js
io.on("connection", (socket) => {
  socket.subscriptions = new Set;

  socket.on("new subscription", (topic) => {
    socket.subscriptions.add(topic);
  });

  // and then
  socket.emit("new message");
  // which is equivalent to
  io.to(socket.id).emit("new message");
});
```

**Reply by scalemaildev** [reply]
That will work perfectly thanks.
